orchagent/routeorch.cpp

```diff
@@ -373,187 +376,318 @@ void RouteOrch::doTask(Consumer& consumer)
+                        it = consumer.m_toSync.erase(it);
+                    }
+                    /* skip prefix which is linklocal or multicast */
+                    else if (ip_prefix.getIp().getAddrScope() != IpAddress::GLOBAL_SCOPE)
+                    {
+                        it = consumer.m_toSync.erase(it);
+                    }
+                    /* fullmask subnet route is same as ip2me route */
+                    else if (ip_prefix.isFullMask() && m_intfsOrch->isPrefixSubnet(ip_prefix, alsv[0]))
+                    {
+                        it = consumer.m_toSync.erase(it);
+                    }
+                    /* subnet route, vrf leaked route, etc */
+                    else
+                    {
+                        if (addRoute(ctx, vrf_id, ip_prefix, nhg))
+                             it = consumer.m_toSync.erase(it);
+                        else
+                            it++;
+                    }
+                }
+                else if (m_syncdRoutes.find(vrf_id) == m_syncdRoutes.end() ||
+                    m_syncdRoutes.at(vrf_id).find(ip_prefix) == m_syncdRoutes.at(vrf_id).end() ||
+                    m_syncdRoutes.at(vrf_id).at(ip_prefix) != nhg)
+                {
+                    if (addRoute(ctx, vrf_id, ip_prefix, nhg))
+                        it = consumer.m_toSync.erase(it);
+                    else
+                        it++;
+                }
+                else
+                    /* Duplicate entry */
+                    it = consumer.m_toSync.erase(it);
 
-            for (auto alias : alsv)
-            {
-                if (alias == "eth0" || alias == "lo" || alias == "docker0")
+                // If already exhaust the nexthop groups, and there are pending removing routes in bulker,
+                // flush the bulker and possibly collect some released nexthop groups
+                if (m_nextHopGroupCount >= m_maxNextHopGroupCount && gRouteBulker.removing_entries_count() > 0)
                 {
-                    excp_intfs_flag = true;
                     break;
                 }
             }
-
-            // TODO: cannot trust m_portsOrch->getPortIdByAlias because sometimes alias is empty
-            if (excp_intfs_flag)
+            else if (op == DEL_COMMAND)
             {
-                /* If any existing routes are updated to point to the
-                 * above interfaces, remove them from the ASIC. */
-                if (removeRoute(vrf_id, ip_prefix))
+                if (removeRoute(ctx, vrf_id, ip_prefix))
                     it = consumer.m_toSync.erase(it);
                 else
                     it++;
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Unknown operation type %s\n", op.c_str());
+                it = consumer.m_toSync.erase(it);
+            }
+        }
+
+        // Flush the route bulker, so routes will be written to syncd and ASIC
+        gRouteBulker.flush();
```

Comment:
two questions:
1 when to flush, it depends on how many entries in consumer.m_toSync ?
2 I saw in flush , the order is removing_entries - > creating_entries -> setting_entries, how do you make sure there is no dependence among them.

e.g , for non-bulk case, the order is : add add del set del set del add add, if bulking, it is {add add add add} {set set} {del del del} , need to make sure if any add/set/del must be before others. We need the experiment data.

Reply:
1 when to flush, it depends on how many entries in consumer.m_toSync ?

Currently there are two trigger points.
- `doTask()` will go through all the `m_toSync` items, after that will trigger one router bulker flush. This is to simplify implementation so no need to remember too much history states
- If there is not more nexthopgroup, and there is some remove_entry queued in the route bulker, will trigger one flush in hope of collecting some nexthopgroup.

2. how do you make sure there is no dependence among them
This is a design choice and after it is fixed, I explicitly merge operation during `create_entry` `remove_entry` and `set_entry_attribute` according this is the sequence.



---
In reply to: [408579209](https://github.com/Azure/sonic-swss/pull/1238#discussion_r408579209) [](ancestors = 408579209)